File: QolmsYappli/QolmsYappli/Models/PartialViewModel/HealthAgeTransitionAreaPartialViewModel.vb
Imports System.Collections.ObjectModel

<Serializable()>
Public NotInheritable Class HealthAgeTransitionAreaPartialViewModel
    Inherits QyPartialViewModelBase(Of HealthAgeViewModel)

#Region "Variable"

    Private _hasData As Boolean = False

    Private _ageData As New List(Of Tuple(Of String, String))()

#End Region

#Region "Public Property"

    Public ReadOnly Property HasData As Boolean

        Get
            Return Me._hasData
        End Get

    End Property

    Public ReadOnly Property AgeData As ReadOnlyCollection(Of Tuple(Of String, String))

        Get
            Return Me._ageData.ToList().AsReadOnly()
        End Get

    End Property

#End Region

#Region "Private Method"

    Private Sub InitializeBy(items As IEnumerable(Of HealthAgeValueItem))

        Me._hasData = False
        Me._ageData = New List(Of Tuple(Of String, String))()

        If items IsNot Nothing Then
            Dim ageN As List(Of HealthAgeValueItem) = items.Where(Function(i) i.HealthAgeValueType = QyHealthAgeValueTypeEnum.Calculation).OrderByDescending(Function(i) i.RecordDate).Take(3).ToList()

            Select Case ageN.Count
                Case 2
                    ageN.Add(New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue})

                Case 1
                    ageN.AddRange(
                        {
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue}
                        }
                    )

                Case 0
                    ageN.AddRange(
                        {
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue}
                        }
                    )

            End Select

            ageN.Reverse()

            For a As Integer = 0 To 2
                If ageN(a).RecordDate <> Date.MinValue And ageN(a).Value > Decimal.Zero Then
                    Me._ageData.Add(New Tuple(Of String, String)(ageN(a).RecordDate.ToString("yyyy年M月d日"), ageN(a).Value.ToString("0.#")))

                    Me._hasData = True
                Else
                    Me._ageData.Add(New Tuple(Of String, String)("－", "－"))
                End If
            Next
        Else
            Me._ageData.Add(New Tuple(Of String, String)("－", "－"))
            Me._ageData.Add(New Tuple(Of String, String)("－", "－"))
            Me._ageData.Add(New Tuple(Of String, String)("－", "－"))
        End If

    End Sub

#End Region

#Region "Constructor"

    ''' <summary>
    ''' <see cref="HealthAgeTransitionAreaPartialViewModel" /> クラスの新しいインスタンスを初期化します。
    ''' </summary>
    ''' <remarks></remarks>
    Protected Sub New()

        MyBase.New()

    End Sub

    ''' <summary>
    ''' <see cref="HealthAgeTransitionAreaPartialViewModel" /> クラスの新しいインスタンスを初期化します。
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub New(model As HealthAgeViewModel)

        MyBase.New(model)

        Me.InitializeBy(model.HealthAgeN)

    End Sub

#End Region

End Class

File: QolmsJotoWebView/Models/PartialViewModel/HealthAgeTransitionAreaPartialViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace MGF.QOLMS.QolmsJotoWebView
{
    [Serializable]
    public sealed class HealthAgeTransitionAreaPartialViewModel : QjPartialViewModelBase<HealthAgeViewModel>
    {
        #region Variable

        private bool _hasData = false;

        private List<Tuple<string, string>> _ageData = new List<Tuple<string, string>>();

        #endregion

        #region Public Property

        public bool HasData
        {
            get { return this._hasData; }
        }

        public ReadOnlyCollection<Tuple<string, string>> AgeData
        {
            get { return this._ageData.ToList().AsReadOnly(); }
        }

        #endregion

        #region Private Method

        private void InitializeBy(IEnumerable<HealthAgeValueItem> items)
        {
            this._hasData = false;
            this._ageData = new List<Tuple<string, string>>();

            if (items != null)
            {
                List<HealthAgeValueItem> ageN = items
                    .Where(i => i.HealthAgeValueType == QjHealthAgeValueTypeEnum.Calculation)
                    .OrderByDescending(i => i.RecordDate)
                    .Take(3)
                    .ToList();

                switch (ageN.Count)
                {
                    case 2:
                        ageN.Add(new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue });
                        break;

                    case 1:
                        ageN.AddRange(new[]
                        {
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue }
                        });
                        break;

                    case 0:
                        ageN.AddRange(new[]
                        {
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue }
                        });
                        break;
                }

                ageN.Reverse();

                for (int a = 0; a <= 2; a++)
                {
                    if (ageN[a].RecordDate != DateTime.MinValue && ageN[a].Value > decimal.Zero)
                    {
                        this._ageData.Add(new Tuple<string, string>(ageN[a].RecordDate.ToString("yyyy年M月d日"), ageN[a].Value.ToString("0.#")));

                        this._hasData = true;
                    }
                    else
                    {
                        this._ageData.Add(new Tuple<string, string>("－", "－"));
                    }
                }
            }
            else
            {
                this._ageData.Add(new Tuple<string, string>("－", "－"));
                this._ageData.Add(new Tuple<string, string>("－", "－"));
                this._ageData.Add(new Tuple<string, string>("－", "－"));
            }
        }

        #endregion

        #region Constructor

        /// <summary>
        /// <see cref="HealthAgeTransitionAreaPartialViewModel" /> クラスの新しいインスタンスを初期化します。
        /// </summary>
        protected HealthAgeTransitionAreaPartialViewModel()
            : base()
        {
        }

        /// <summary>
        /// <see cref="HealthAgeTransitionAreaPartialViewModel" /> クラスの新しいインスタンスを初期化します。
        /// </summary>
        public HealthAgeTransitionAreaPartialViewModel(HealthAgeViewModel model)
            : base(model)
        {
            this.InitializeBy(model.HealthAgeN);
        }

        #endregion
    }
}
